### ISSUE TITLE
assembler label problem and work on MPIR

### ISSUE THREAD

**Issue by GitMensch** [LADSoft/OrangeC#330]
The following assembler constructs don't compile (generated by MPIR [configure and make otherwise works] but most of the assembler sources look that way so all sub-directories fail to be generated):

```asm
	.text
	.align	4, 0x90
	.globl	___gmpn_add_err1_n
	.type	___gmpn_add_err1_n,@function
___gmpn_add_err1_n:

	mov	8(%rsp), %rax    

	push	%rbx
	push	%rbp
	push	%r12

	xor %rbx, %rbx              
	xor %rbp, %rbp
	lea	(%rdi,%r9,8), %rdi     
	lea	(%rsi,%r9,8), %rsi
	lea	(%rdx,%r9,8), %rdx

```

>Error tmp-add_err1_n.s(74): Label '___gmpn_add_err1_n' already exists.
Error tmp-add_err1_n.s(76): Constant value expected
Error tmp-add_err1_n.s(78): Unknown token sequence
Error tmp-add_err1_n.s(79): Unknown token sequence
Error tmp-add_err1_n.s(80): Unknown token sequence
Error tmp-add_err1_n.s(82): Unknown token sequence
Error tmp-add_err1_n.s(83): Unknown token sequence
Error tmp-add_err1_n.s(84): Constant value expected
Error tmp-add_err1_n.s(85): Constant value expected
Error tmp-add_err1_n.s(86): Constant value expected

Complete file (don't ask me why there are so much empty lines in it...):
[tmp-add_err1_n.s](https://github.com/LADSoft/OrangeC/files/2795162/tmp-add_err1_n.s.txt)

**Comment by LADSoft** [comment]
ok there are four problems - I'm mostly writing this now so you will be aware of #4...

1) .globl isn't working as expected - will fix
2) uses gnu assembler syntax (register names and operand order) - maybe I will add a command line switch to accept GNU syntax?
3) this is 64 bit code - maybe I will add a command line switch to select that as well (normal usage for this assembler is to embed a directive in the source file)
4) because it is 64-bit code and OCC is a 32-bit compiler, when we get the output compiling it still won't run...

**Comment by LADSoft** [comment]
have not yet gotten traction on this...    maybe this week

**Comment by LADSoft** [comment]
i'm starting to look at this...   the .globl problem is probably due to the .type statement, so I guess a part of getting the end result we want has got to be to completely address #236 as well...

**Comment by GitMensch** [comment]
> completely address #236

8-|
... completely sounds marvelous (and like a lot of work)

As there's a lot of code out there that actually is using gas I think it is worth the work.

**Comment by LADSoft** [comment]
sorry for the delay on this and the other things you've brought up recently...  I've been dealing with a lot of unexpected compiler bugs that surfaced when I replaced pointers with 'unique_ptr'.    To get the unique_ptr to work I had to fix a few bugs and it kind of snowballed on me...   fix for one bug broke three other things sigh.    I was trying to work it as a background task on another branch but the code diverged so much i decided to just get it finished.

As of tonite I have the compiler compiling itself again well enough that it can compile other programs...    but there is a problem with one of the more problematic programs in the test suite just hanging when I run it.   I Hope to have any remaining problems worked out within the next few days and get back to working on the issues again.

**Comment by GitMensch** [comment]
Good to read you make progress.
There's no nneed to push it, take your time.

**Comment by LADSoft** [comment]
thanks!   

FWIW now I've apparently found some arcane bug in mingw64 that doesn't exist in any of the other compilers, including orange C.    Guess I will work on finding a workaround tomorrow...

**Comment by chuggafan** [comment]
mingw64 bug? That might be because you don't have the latest compiler version (GCC 7.2 or 8 should be available)? If it's up to date and still buggy, I recommend reporting it because it should be just a straight GCC ontop of an emulation layer...

**Comment by LADSoft** [comment]
thanks for the tip.   That helped.   8.10 seems to fix it; Previously I had 7.30.

**Comment by LADSoft** [comment]
i'm gonna backtrack on that.   Seems to work with the i686 builds of mingw64 but not with the x86_64 builds.   Unfortunately we need the latter for appveyor...   so I guess there has to be a workaround at minimum but if I can get that far I'll also report it.

**Comment by LADSoft** [comment]
mea culpa, mingw works fine.   There was a longstanding problem that just became uncovered for reasons unknown.  If orangec was compiled with a 64 bit compiler it then wouldn't emit exception information properly.   Will release a fix for that soon...

**Comment by LADSoft** [comment]
I haven't forgotten this...   I'm going to work on #355 for a a while then I will come back to it :)

**Comment by LADSoft** [comment]
@GitMensch question for you, in the opening post of this issue some source code was given that doesn't compile.

the GAS documentation says that instruction size should be postfixed on the instruction, e.g.

> xorl $4,%eax

but the example doesn't use the size postfix values.   So I'm wondering what the rules for omitting them are.   If omitted does the size default to DWORD?   Or is the size derived from the register sizes 
the way it is done in the intel syntax?

**Comment by LADSoft** [comment]
never mind I found it in the documentation :smile:

**Comment by LADSoft** [comment]
I'm going to check in a fix for this tonite or tomorrow; but it only addresses the mnemonic/operand problems and related issues.    The problem of GAS-like directives will be fixed in relation to #236 (also in this milestone)

**Comment by GitMensch** [comment]
Rechecked with the merge of assembler branch CI build, does not work:

```asm

	.text
	.align	4, 0x90
	.globl	___gmpn_add_err1_n
	.type	___gmpn_add_err1_n,@function
___gmpn_add_err1_n:

	mov	8(%rsp), %rax    

	push	%rbx
	push	%rbp
	push	%r12

	xor %rbx, %rbx              
	xor %rbp, %rbp
	lea	(%rdi,%r9,8), %rdi     
	lea	(%rsi,%r9,8), %rsi
	lea	(%rdx,%r9,8), %rdx

	test	$1, %r9            
	jnz	Lodd

Leven:	
	lea	-8(%r8,%r9,8), %r8   
	neg	%r9                
	jmp	Ltop

	.align	4, 0x90
Lodd:                           
	lea	-16(%r8,%r9,8), %r8     
	neg	%r9                   
	shr	$1, %rax            
	mov	(%rsi,%r9,8), %r12
	adc	(%rdx,%r9,8), %r12
	cmovc	8(%r8), %rbx           
	mov	%r12, (%rdi,%r9,8)
	setc	%al                 
	inc	%r9                   
	jz	Lend              

	.align	4, 0x90
Ltop:
       mov     (%rsi,%r9,8), %r12     
	shr     $1, %rax        
	adc     (%rdx,%r9,8), %r12
	mov     $0, %r11          
	mov     %r12, (%rdi,%r9,8)
	mov     $0, %r10          
	mov     8(%rsi,%r9,8), %r12    
	cmovc   (%r8), %r10        
	adc     8(%rdx,%r9,8), %r12
	cmovc   -8(%r8), %r11      
	setc    %al             
	add     %r10, %rbx          
	adc     $0, %rbp
	add     %r11, %rbx          
	mov     %r12, 8(%rdi,%r9,8)
	adc     $0, %rbp
	add     $2, %r9           
	lea     -16(%r8), %r8     
	jnz     Ltop          

Lend:	
	mov	%rbx, (%rcx)         
	mov	%rbp, 8(%rcx)

	pop	%r12
	pop	%rbp
	pop	%rbx
	ret
	
```

Resulting in

> Error tmp-add_err1_n.s(6): Label '___gmpn_add_err1_n' already exists.
> Error tmp-add_err1_n.s(8): Constant value expected
> Error tmp-add_err1_n.s(10): Unknown token sequence
> ...

Or is this related to the GAS-like directives issue?

**Comment by LADSoft** [comment]
this is a problem with not supporting the gas-like directives, but I'll reopen this until it actually compiles your code snippet...

**Comment by LADSoft** [comment]
Hi @GitMensch, i fixed this so it will assemble your example now.

There may also have been some confusion; you additionally need to specify the command line option --gas:

>      Oasm --gas test.asm

This was listed in the command line help already but I've updated the .MD files as well now.   Note that I had additionally addressed #249 in this milestone, which allows for 'long' names to be used in switches as we move forward.

**Comment by GitMensch** [comment]
Not sure but I guess this is related:

```text
$ oasm --gas test.s.txt
oasm Version 6.0.43.40 Copyright (C) LADSoft 2006-2019
Error test.s.txt(53): Unrecognized opcode
1 Errors, 0 Warnings
```

[test.s.txt](https://github.com/LADSoft/OrangeC/files/3650726/test.s.txt)

**Comment by GitMensch** [comment]
@LADSoft just to rechecked: should I create a new issue for the "Unrecognized opcode" error? [this one is "more" of a show-stopper to the current milestone *for me*]

**Comment by LADSoft** [comment]
sorry I just caught this...   will fix first thing tomorrow evening.    You can leave this and other things in this issue since the are related to the work here and I'll try to be more responsive.

**Comment by GitMensch** [comment]
No problem. Just a note about how all those issues were found:

* download [MPIR 3.0.0 source](http://mpir.org/downloads.html)
* unpack somwhere
* if not already done: download [YASM](https://yasm.tortall.net/Download.html) and place it somewhere
* start any shell (for example MSYS, MSYS2, cygwin, BashOnWindows; MSYS/MSYS2 works best)
* ensure that OrangeC/bin is in path
* execute

```shell
cd /some/path/mpir-3.0.0
mkdir build_occ
cd build_occ
../configure 'CC=occ --nologo' 'CFLAGS=-pa--gas' 'MAKE=omake --nologo' LD='olink --nologo' 'CPP=ocpp --nologo' CPP_FOR_BUILD='ocpp --nologo' 'CXX=occ --nologo' 'CXXCPP=ocpp --nologo' GREP=ogrep EGREP=ogrep "NM=onm --nologo " --with-yasm=/path/to/yasm-1.3.0-win64.exe --disable-static --enable-shared
omake
```
The configure-script generated config.log already showed some asserts, other issues and missing files (it is *not* intended that all tests result in "yes works", but obviously an assert should not be found there). The others were found when running `omake`.

**Comment by LADSoft** [comment]
thanks...

**Comment by LADSoft** [comment]
the only trouble I saw with the config script was that ocpp doesn't have a configuration file so it couldn't find the orangec includes directory automatically...    

I fixed that and the assembler bug you pointed out the other day.   Then after the configure script finished I looked through the config.log for the word 'assert' but couldn't find anything that seemed relevant.   Just a lot of references to the assert header in various code snippets the autoconfig was using to look for supported functions.   What should I be looking for?

When omake ran occ compiled a bunch of c files then got to a file called get_d.c.   The make bailed at that point after complaining about accesses to members of a structure 'ieee_double_extract'.   Occ thinks it was forward declared but not given a body before being used.   Is this something I need to look into?

**Comment by GitMensch** [comment]
The asserts in the config.log are solved with all those recent changes. Last time I've checked make in mpn broke, if have to recheck.
The structure issue sounds like something that needs a fix (builds with gcc,clang,msc).

**Comment by LADSoft** [comment]
the problem is a file in mpn\generic.

looks like some problem with the configure script.   from  build_occ\config.h:

```
/* #undef HAVE_DOUBLE_IEEE_BIG_ENDIAN */
/* #undef HAVE_DOUBLE_IEEE_LITTLE_ENDIAN */
/* #undef HAVE_DOUBLE_IEEE_LITTLE_SWAPPED */

```
so ieee_double_extract isn't getting defined from gmp-impl.h.   the build for VC seems to have the little endian one defined and I think OCC should too.

Will see if I can figure out why the config script fails in this regard but may need help...

**Comment by LADSoft** [comment]
ok I *think* the problem may be that the configure scripts assume the object files are binary in nature and are getting thrown for a loop by the fact the the OCC object files are ascii.

meanwhile I tried #defining the little endian one in config.h but still couldn't get past this...

**Comment by LADSoft** [comment]
i figured out what I was doing wrong in config.h and it at least compiles everything in the MPN directory now.   Then I had to add 'AR=olib' and regenerate the configuration...   right now it is failing on the library insert I think because some of the object files are being generated in what appears to be COFF format...   I think that is possibly because the project uses YASM?

**Comment by LADSoft** [comment]
there is some problem with occ... if I compile occ with MSVC it then compiles the mpn directory, but if I then compile it with itself I then get errors when compiling the mpn directory e.g. in gmp-impl.h.   Will fix soon...

BTW if the HAVE_DOUBLE_IEEE_LITTLE_ENDIAN detection and similar things that parse a binary formatted object file is a big deal it looks like I have already written a 'binary' version of the object file handling, but it is probably out of date and needs to be vetted further before it can be used...

**Comment by GitMensch** [comment]
> there is some problem with occ... if I compile occ with MSVC it then compiles the mpn directory, but if I then compile it with itself I then get errors when compiling the mpn directory e.g. in gmp-impl.h. Will fix soon...

Sounds good. So I'll wait until I hear something again (but effectively I'll just do the steps above and when those work run `omake check` - things you could do yourself, too). If there's a working binary version then I think occ and oasm proved themself to be usable for a lot of other projects :-)

> BTW if the HAVE_DOUBLE_IEEE_LITTLE_ENDIAN detection and similar things that parse a binary formatted object file is a big deal [...]

For now the effect should only be that MPIR will do "normal" double/float handling and no binary optimizations.

> it looks like I have already written a 'binary' version of the object file handling, but it is probably out of date and needs to be vetted further before it can be used...

Interesting. I think it would be reasonable to handle this with a new issue. It will be interesting to see if/what the effect in case of necessary compile/link time (and object file's size) this has.

**Comment by LADSoft** [comment]
ok so there is some problem with attribute((noreturn)) in versions compiled with itself but for now it seems you can get around it by making sure to autoconfigure against the version of the compiler you intend to use.   Autoconfigure apparently detects it isn't going to work and causes the compile not to use it.   I will fix it tomorrow...

MPIR won't compile if you don't have one of the HAVE_DOUBLE flags set.   In that case it doesn't define a needed union, so the compile throws errors.   I've been assuming we need to get one defined and the way to get one defined is to switch to a binary object file format and then the autoconfig will detect floating point.

I'm with you, the binary object format needs a new issue...   FWIW I originally wrote that in hopes it will speed things up but didn't see much happening and then decided not to finish it out.   That was back when the compiler had an independent version of the object file format handling though.   Maybe with the longer builds we will see something...   it should cut the files size by quite a bit because right now every literal byte of data is encoded as two ASCII characters...

That won't fix everything though, use of YASM pretty much dictates it won't link.   And meanwhile 'omake check' is faililng, it doesn't know how to create ../libmpir.la.

Will pick it up tomorrow :smile:

**Comment by GitMensch** [comment]
I think there's a "slightly" topic drift here ;-)

> MPIR won't compile if you don't have one of the HAVE_DOUBLE flags set. In that case it doesn't define a needed union, so the compile throws errors.

If this is the case then it is a MPIR bug, which is possible as most (all other?) compilers have a binary format. The solution is *not* to have a binary object format but to create and apply a minimal patch to the MPIR code - and/or to check if https://github.com/wbhart/mpir already contains the fix - you can get artifacts for it very likely from https://ci.appveyor.com/project/wbhart/mpir and in the meanwhile from https://ci.appveyor.com/project/GitMensch/mpir ...

**Comment by LADSoft** [comment]
I fixed the noreturn problem.   It was much deeper than I first suspected, I'm surprised the compiler worked as well as it did...

**Comment by LADSoft** [comment]
@GitMensch - 

the config scripts assume MSVC when it isn't one of the GCC variants, and want to use LIB.EXE.   There is even a line in there that says something to the effect of ,FIXME, WE REALLY SHOULD LET THE USER SPECIFY THE LIBRARIAN.,

a more serious issue is that YASM creates object files the occ toolchain cannot process, so even if I run the librarian separately I get a bunch of errors about object files having a bad format.

I will look at the 'omake check' issue next...

**Comment by GitMensch** [comment]
For config scripts: just workaround as with GnuCOBOL: directly after configure adjust the generated libtool file to use correct librarian.
For yasm: no idea without a computer at hand ;-)

**Comment by LADSoft** [comment]
what I'm getting from configure is assembly language files that cannot be processed by oasm in either mode; hence YASM.    But if we need YASM to create object files it just isn't going to work.

**Comment by LADSoft** [comment]
I was thinking, if there is some way to get autoconfig to generate gas-compatible assembly language files from the existing stubs, then we wouldn't need YASM and we could also test the oasm GAS compatibility some more.    But I haven't a clue as to whether it is even possible, or how to go about it if it is...

**Comment by GitMensch** [comment]
You may want to ask on the mpir mailing list if there's any option to not use yasm on x86/64 (it isn't used for other cpus).

**Comment by LADSoft** [comment]
i asked and according to the author of MPIR it can't use GAS.      yasm is a requirement at the moment as the MPIR code uses features specific to yasm.   there have been proposals to rewrite the assembly code for GAS but noone has found the time to do it.

I've gone as far as I can with MPIR.

**Comment by GitMensch** [comment]
I think there is at least one thing to do here as there is one assembler related error with MPIR (and a bunch of asserts) [I don't expect the result to "work" any more because of the yasm+linking isuse].

Things to do: setup system (using current "MINGW64, then static lib") as noted [above](https://github.com/LADSoft/OrangeC/issues/330#issuecomment-535885062).

Shown assembler issues (just reran `omake -k` after the first time) - @LADSoft please reopen because of these:

Syntax errors:

~~~
Error tmp-divexact_1.s(107): Label '%' already exists.
Error tmp-modexact_1c_odd.s(113): Label '%' already exists.
~~~

Current assert (3 times):

~~~
Access Violation:(C:\dev\OrangeC\bin\oasm.exe)
CS:EIP 0023:00506A3D  SS:ESP 002B:0019E58C
EAX: 0000001C  EBX: 00000000  ECX: 0000001C  EDX: 00000000  flags: 00010202
EBP: 005B172E  ESI: 00000018  EDI: 00000018
 DS:     002B   ES:     002B   FS:     0053   GS:     002B


CS:EIP  0F B6 00 83 E0 01 0F 95 C0 22 C0 74 11 8B C6 83
~~~

Note: one could either ignore the error messages that result from the [missing MPIR structure definition](https://github.com/wbhart/mpir/issues/278) or adjust gmp-impl.h to always use the little-endian definition (which I've locally did with changing the `#if ...` sequence to `#if .. #elif ... #else`).

**Comment by LADSoft** [comment]
now it compiles a bunch of stuff; when it gets to DOCs it does this:

>        restore=: && backupdir=".am$$$" && am__cwd=`pwd` && CDPATH="${ZSH_VERSION+.}:" && cd ../../doc && rm -rf $backupdir && mkdir $backupdir && if (/bin/sh /c/dev/mpir-3.0.0/missing makeinfo --version) >/dev/null 2>&1; then for f in ../../doc/mpir.info ../../doc/mpir.info-[0-9] ../../doc/mpir.info-[0-9][0-9] .i[0-9] .i[0-9][0-9]; do if test -f $f; then mv $f $backupdir; restore=mv; else :; fi; done; else :; fi && cd "$am__cwd" ; if /bin/sh /c/dev/mpir-3.0.0/missing makeinfo -I ../../doc -o ../../doc/mpir.info ../../doc/mpir.texi; then rc=0; CDPATH="${ZSH_VERSION+.}:" && cd ../../doc; else rc=$?; CDPATH="${ZSH_VERSION+.}:" && cd ../../doc && $restore $backupdir/* `echo "./../../doc/mpir.info" | sed 's|[^/]*$||' `; fi; rm -rf $backupdir; exit $rc
> /c/dev/mpir-3.0.0/missing: line 81: makeinfo: command not found
> WARNING: 'makeinfo' is missing on your system.
>          You should only need it if you modified a '.texi' file, or
>          any other file indirectly affecting the aspect of the manual.
>          You might want to install the Texinfo package:
>          <https://www.gnu.org/software/texinfo/>
>          The spurious makeinfo call might also be the consequence of
>          using a buggy 'make' (AIX, DU, IRIX), in which case you might
>          want to install GNU make:
>          <https://www.gnu.org/software/make/>
> 
>

**Comment by GitMensch** [comment]
As noted this should not happen, and may be an `omake` issue. In any case you should be able to "skip" this directory by either running `omake` in the subdirectories excluding "doc" or using `omake -k`.

**Comment by LADSoft** [comment]
the only reason i was holding this open is because of my comment that the docs wouldn't build for mpir, and the resulting comment from @GitMensch that maybe there was a bug in omake.    

I was able to verify tonite that right now the mpir docs directory does compile.   I think at some point I figured out how to install makeinfo?   The rest of the project compiles right up to the point where it tries to link the DLL.   Since this is likely due to the usage of yasm I'm going to close this.
